File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/UnknownAttributeInfo.kt
package com.jetbrains.signatureverifier.serialization

import org.bouncycastle.asn1.DLSequence
import org.bouncycastle.asn1.cms.Attribute

data class UnknownAttributeInfo(
  val identifier: StringInfo,
  val content: ByteArray
) : AttributeValueInfo() {

  override fun toAttributeDLSequence(): DLSequence = DLSequence.getInstance(content) as DLSequence

  constructor(attribute: Attribute) : this(
    StringInfo.getInstance(attribute.attrType),
    attribute.getEncoded("DER")
  )
}

File: net/JetBrains.Serialization/src/Dataholders/AttributeInfos/UnknownAttributeInfo.cs
using Newtonsoft.Json;
using Org.BouncyCastle.Asn1;
using Attribute = Org.BouncyCastle.Asn1.Cms.Attribute;

namespace JetBrains.Serialization;

[JsonObject(MemberSerialization.OptIn)]
public class UnknownAttributeInfo : AttributeInfo
{
  [JsonProperty("Identifier")] protected override TextualInfo Identifier { get; }

  [JsonProperty("Content")] private IEncodableInfo _content;

  [JsonConstructor]
  private UnknownAttributeInfo(TextualInfo identifier, IEncodableInfo content)
  {
    Identifier = identifier;
    _content = content;
  }

  public UnknownAttributeInfo(Attribute attribute)
    : this(TextualInfo.GetInstance(attribute.AttrType), attribute.AttrValues.ToEncodableInfo())
  {
  }

  public override Asn1Encodable GetPrimitiveContent() => _content.ToPrimitive();
}